src/kernel/pgtbl.c

```diff
@@ -4,184 +4,36 @@
-	 * we won't use it for other kernel objects. */
-	if (unlikely(cos_cas((unsigned long *)pte, orig_v, new_v) != CAS_SUCCESS)) {
-		/* restore the ref cnt. */
-		retypetbl_deref((void *)(orig_v & PGTBL_FRAME_MASK));
-		return -ECASFAIL;
-	}
-	/* Return the pte ptr, so that we can release the page if the
-	 * kobj activation failed later. */
-	*pte_ret = (unsigned long *)pte;
-
-	return 0;
+    return chal_pgtbl_kmem_act(pt, addr, kern_addr, pte_ret);
 }
 
 /* Return 1 if quiescent past since input timestamp. 0 if not. */
 int
 tlb_quiescence_check(u64_t timestamp)
 {
-	int i, quiescent = 1;
-
-	/* Did timer interrupt (which does tlb flush
-	 * periodically) happen after unmap? The periodic
-	 * flush happens on all cpus, thus only need to check
-	 * the time stamp of the current core for that case
-	 * (assuming consistent time stamp counters). */
-	if (timestamp > tlb_quiescence[get_cpuid()].last_periodic_flush) {
-		/* If no periodic flush done yet, did the
-		 * mandatory flush happen on all cores? */
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			if (timestamp > tlb_quiescence[i].last_mandatory_flush) {
-				/* no go */
-				quiescent = 0;
-				break;
-			}
-		}
-	}
-	if (quiescent == 0) {
-		printk("from cpu %d, t %llu: cpu %d last mandatory flush: %llu\n", get_cpuid(), timestamp, i,
-		       tlb_quiescence[i].last_mandatory_flush);
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			printk("cpu %d: flush %llu\n", i, tlb_quiescence[i].last_mandatory_flush);
-		}
-	}
-
-	return quiescent;
+    return chal_tlb_quiescence_check(timestamp);
 }
 
 int
 cap_memactivate(struct captbl *ct, struct cap_pgtbl *pt, capid_t frame_cap, capid_t dest_pt, vaddr_t vaddr)
 {
-	unsigned long *    pte, cosframe, orig_v;
-	struct cap_header *dest_pt_h;
-	u32_t              flags;
-	int                ret;
-
-	if (unlikely(pt->lvl || (pt->refcnt_flags & CAP_MEM_FROZEN_FLAG))) return -EINVAL;
-
-	dest_pt_h = captbl_lkup(ct, dest_pt);
-	if (dest_pt_h->type != CAP_PGTBL) return -EINVAL;
-	if (((struct cap_pgtbl *)dest_pt_h)->lvl) return -EINVAL;
-
-	pte = pgtbl_lkup_pte(pt->pgtbl, frame_cap, &flags);
-	if (!pte) return -EINVAL;
-	orig_v = *pte;
-
-	if (!(orig_v & PGTBL_COSFRAME) || (orig_v & PGTBL_COSKMEM)) return -EPERM;
-
-	assert(!(orig_v & PGTBL_QUIESCENCE));
-	cosframe = orig_v & PGTBL_FRAME_MASK;
-
-	ret = pgtbl_mapping_add(((struct cap_pgtbl *)dest_pt_h)->pgtbl, vaddr, cosframe, PGTBL_USER_DEF);
-
-	return ret;
+    return chal_cap_memactivate(ct, pt, frame_cap, dest_pt, vaddr);
 }
 
 int
 pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, pgtbl_t pgtbl, u32_t lvl)
 {
-	struct cap_pgtbl *pt;
-	int               ret;
-
-	pt = (struct cap_pgtbl *)__cap_capactivate_pre(t, cap, capin, CAP_PGTBL, &ret);
-	if (unlikely(!pt)) return ret;
-	pt->pgtbl = pgtbl;
-
-	pt->refcnt_flags = 1;
-	pt->parent       = NULL; /* new cap has no parent. only copied cap has. */
-	pt->lvl          = lvl;
-	__cap_capactivate_post(&pt->h, CAP_PGTBL);
-
-	return 0;
```

Comment:
Same confusion as above.
Even the pgtbl operation depends on platform, should the protocol with ```*capactivate_pre``` and ```*capactivate_post``` be in a higher-level function (in the kernel instead of chal)?

Reply:
The parameter list of the containing function is going to be different.

Reply:
I did not get it.
Which function's parameter list is changed? ```pgtbl_activate```, ```cap_capactivate_*``` or ```chal_pgtbl_activate```?
Could you change function ```pgtbl_activate``` to
```
cap_capactivate_pre();
chal_pgtbl_activate();
cap_capactivate_post();
```
instead of putting everything in ```chal_pgtbl_activate```?

Reply:
the pgtbl_activate itself. This modification is not final, and there is going to be more modifications that follows.